### ISSUE TITLE
45GS02 compilation error

### ISSUE THREAD

**Issue by M3wP** [hth313/Calypsi-tool-chains#42]
When compiled with "cc6502 --core 45gs02 -S -o main.o main.c", the following code results in incorrect assembly.  The high byte of the r value is being overwritten by the low byte.  I tried ensuring every element of the statement was cast to unsigned short but that did not help.

int main() {
    unsigned char* vr = (unsigned char*)0xD052;
    unsigned short r = 0;

    r = ((0x03 & (unsigned short)vr[1]) << 8) | (unsigned short)vr[0];

    return 0;
}

**Comment by hth313** [comment]
This is caused by a problem in the code generator. There is a  code generator rule for this pattern of `or`-ing together two bytes that violates a register allocation rule contract. This can cause it to reuse the same register for both bytes.

I have revised the rule, actually replaced it with a couple of variant rules that does it right and also should give slightly better code, the changed compiler generates:

```
0191                r = ((0x03 & (unsigned short)vr[1]) << 8) | (unsigned short)vr[0];
    \ 0010 a903                 lda     #3
    \ 0012 a301                 ldz     #1
    \ 0014 32..                 and     (_Zp+2),z
    \ 0016 85..                 sta     zp:_Zp+1
    \ 0018 a300                 ldz     #0
    \ 001a b2..                 lda     (_Zp+2),z
    \ 001c 85..                 sta     zp:_Zp
```

Do you need a fixed compiler urgently or are you able to work around the problem? In that case, what host platform do you use?

**Comment by M3wP** [comment]
I suppose its not urgent.  I'm using Windows.

**Comment by hth313** [comment]
This is fixed in the 5.5 release that is now available.

**Comment by hth313** [comment]
Closed for now as fixed.
